core/src/KokkosExp_MDRangePolicy.hpp

```diff
@@ -113,9 +113,24 @@ struct Rank {
 };
 
 namespace Impl {
+// Checked narrowing conversion that calls abort is the cast changes the value
+template <class To, class From>
+constexpr To checked_narrow_cast(From arg) {
+  constexpr const bool is_different_signedness =
+      (std::is_signed<To>::value != std::is_signed<From>::value);
+  auto const ret = static_cast<To>(arg);
+  if (static_cast<From>(ret) != arg ||
+      (is_different_signedness && (arg < From{}) != (ret < To{}))) {
+    Kokkos::abort("unsafe narrowing conversion");
+  }
+  return ret;
+}
 // NOTE prefer C array U[M] to std::initalizer_list<U> so that the number of
 // elements can be deduced (https://stackoverflow.com/q/40241370)
-template <class Array, class U, std::size_t M>
+// NOTE for some unfortunate reason the policy bounds are stored as signed
```

Comment:
This note needs to be expanded upon.  The problem isn't just signed vs. unsigned numbers, it is really because there are possibly three separate element types involved, one of which may have a different sign than the others:  `IndexType`, `Array::value_type` and `U`.

Even after reading the note, it was not obvious why `IndexType` is here at all, or why we cannot convert to an array of `IndexType` here.